dialogue-core/src/main/java/com/palantir/dialogue/core/BalancedScoreTracker.java

```diff
@@ -0,0 +1,331 @@
+ * This is intended to be a strict improvement over Round Robin and Random Selection which can leave fast servers
+ * underutilized, as it sends the same number to both a slow and fast node. It is *not* appropriate for transactional
+ * workloads (where n requests must all land on the same server) or scenarios where cache warming is very important.
+ * {@link PinUntilErrorNodeSelectionStrategyChannel} remains the best choice for these.
+ */
+final class BalancedScoreTracker {
+    private static final Logger log = LoggerFactory.getLogger(BalancedScoreTracker.class);
+
+    private static final Comparator<ScoreSnapshot> BY_SCORE = Comparator.comparingInt(ScoreSnapshot::getScore);
+    private static final Duration FAILURE_MEMORY = Duration.ofSeconds(30);
+    private static final double FAILURE_WEIGHT = 10;
+
+    private final ImmutableList<ChannelScoreInfo> channelStats;
+    private final Random random;
+    private final Ticker clock;
+
+    BalancedScoreTracker(
+            int channelCount, Random random, Ticker ticker, TaggedMetricRegistry taggedMetrics, String channelName) {
+        Preconditions.checkState(channelCount >= 2, "At least two channels required");
+        this.random = random;
+        this.clock = ticker;
+        this.channelStats = IntStream.range(0, channelCount)
+                .mapToObj(index -> new ChannelScoreInfo(
+                        index, clock, PerHostObservability.create(channelCount, taggedMetrics, channelName, index)))
+                .collect(ImmutableList.toImmutableList());
+
+        registerGauges(taggedMetrics, channelName, channelStats);
+    }
+
+    /**
+     * Returns all channels, ordered with the best score first. Called on every request, so needs to be performant!
+     * Callers *must* use the {@link ChannelScoreInfo#startRequest} and {@link ChannelScoreInfo#onSuccess} etc
+     * methods to feed information back into the tracker.
+     */
+    public ChannelScoreInfo[] getChannelsByScore() {
+        // pre-shuffling is pretty important here, otherwise when there are no requests in flight, we'd
+        // *always* prefer the first channel of the list, leading to a higher overall load.
+        List<ChannelScoreInfo> shuffledMutableStats = shuffleImmutableList(channelStats, random);
+
+        ScoreSnapshot[] snapshotArray = new ScoreSnapshot[shuffledMutableStats.size()];
+        for (int i = 0; i < snapshotArray.length; i++) {
+            snapshotArray[i] = shuffledMutableStats.get(i).computeScoreSnapshot();
+        }
+
+        Arrays.sort(snapshotArray, BY_SCORE);
+
+        ChannelScoreInfo[] returnArray = new ChannelScoreInfo[snapshotArray.length];
+        for (int i = 0; i < snapshotArray.length; i++) {
+            returnArray[i] = snapshotArray[i].delegate;
+        }
+        return returnArray;
+    }
+
+    public ChannelScoreInfo getSingleBestChannelByScore() {
+        // TODO(dfox): in theory we could optimize this by just looping manually and keeping track of the max
```

Comment:
Would we need to break ties in random order to prevent affinity? I think we considered this when we implemented the balanced channel originally.

Reply:
I see, avoiding the ChannelScoreInfo and operate directly on ScoreSnapshots